Sega Genesis controllers update

KOPRajs opened the pull request:
The goal is to add a 3 button joypad, the _Team Player_ and the _4 Way Play_ controllers for Sega Genesis.

garbear commented on addons/game.controller.genesis.3button/addon.xml:
```diff
@@ -0,0 +1,16 @@
+            <icon>resources/icon.png</icon>
+        </assets>
+        <summary lang="en_GB">Sega Genesis controller</summary>
+        <description lang="en_GB">The original controller for the Sega Genesis was released in 1988. It had three main face buttons. Sega released a six-button version in 1993.</description>
```
Good idea using the same description, it applies perfectly and 1 less string for translators.

garbear commented on addons/game.controller.genesis.6button/addon.xml:
```diff
@@ -1,7 +1,7 @@
 <addon id="game.controller.genesis.6button"
-        name="Sega Genesis Controller"
-        version="1.0.22"
+        name="Genesis Controller (6-Button)"
```
The summary here and in the strings file also should be updated.

garbear commented:
Looks like this was fundamentally done right, and good catch on the other fixes. What should we do for the 3-button controller image?

garbear commented:
Can you tag me as a reviewer in the future or CC me in the description? I'm gonna start checking my Kodi notifications a bit less and might miss a game-related PR or issue.

KOPRajs commented:
> What should we do for the 3-button controller image?

How about this? As I've said in the forum, the source picture can be found across the internet on more than 20 different sites, so I don't know who is the original author. I've removed the branding, added transparency, added the shadow and added the background to match the other controller.

> Can you tag me as a reviewer in the future or CC me in the description?

I will.

> I'm gonna start checking my Kodi notifications a bit less and might miss a game-related PR or issue.

Pitty to hear that you're gonna have less time for Retroplayer, but I can understand that very well.

You can merge this when you think it is ready, I will create another PR for the other 2 controllers when I get some time.

KOPRajs commented:
While testing it, I've found out that I need to make a few more modifications.

KOPRajs commented:
@garbear Can you please give me a hand here?

Is there some kind of a default mapping from the Default controller to every other controller? I mean, when you setup _feature name="rightbumper"_ for the Default controller, the Sega emulator knows that it is _feature name="c"_ for the 6 button controller.

I can't find anything like a translation from _game.controller.default_ to _game.controller.genesis.6button_.

garbear commented:
> Is there some kind of mapping from the Default controller to every other controller?

Yes, at the very end, we fall back to C++.

Search https://github.com/kodi-game/game.libretro/tree/Nexus/src/input for the tables the define the default fallbacks for the known controllers.

garbear commented on addons/game.controller.genesis.3button/addon.xml:
```diff
@@ -0,0 +1,15 @@
+        <assets>
+            <icon>resources/icon.png</icon>
+        </assets>
+        <summary lang="en_GB">Sega Genesis controller (3-Button)</summary>
```
Button should probably be "Sentence case." for summary

KOPRajs replied:
I was trying to follow the pattern of other controllers here:
https://github.com/kodi-game/controller-topology-project/blob/1a90e1241206b564d0fad7bb64a2091729ce9d30/addons/game.controller.joystick.2button/addon.xml#L15

So should I keep it? Or do you want me to change this one and you will fix the rest? Or do we make another PR to fix this for all controllers at once?

KOPRajs replied:
Other candidates:
https://github.com/kodi-game/controller-topology-project/blob/1a90e1241206b564d0fad7bb64a2091729ce9d30/addons/game.controller.joystick.4button/resources/language/resource.language.en_gb/strings.po#L20
https://github.com/kodi-game/controller-topology-project/blob/1a90e1241206b564d0fad7bb64a2091729ce9d30/addons/game.controller.konami.justifier.player2/resources/language/resource.language.en_gb/strings.po#L20
https://github.com/kodi-game/controller-topology-project/blob/1a90e1241206b564d0fad7bb64a2091729ce9d30/addons/game.controller.ps.guncon.western/resources/language/resource.language.en_gb/strings.po#L20
https://github.com/kodi-game/controller-topology-project/blob/1a90e1241206b564d0fad7bb64a2091729ce9d30/addons/game.controller.saturn.3d.western/resources/language/resource.language.en_gb/strings.po#L20
https://github.com/kodi-game/controller-topology-project/blob/1a90e1241206b564d0fad7bb64a2091729ce9d30/addons/game.controller.saturn.virtua.gun.eu/resources/language/resource.language.en_gb/strings.po#L20

So I'd keep it.

garbear replied:
Generally, a "name" field should be "Title Case" and "summary" and "description" fields should be "Sentence case." However, "name" isn't translated, while summary is, so I think we use summary as a name in the GUI. In that case, summary should also be title case.

I think the question is, what looks better in the UI. Check the port dialog and controller mapper, and if you think that Title Case looks better than Sentence case then we'll move everything to that. I suspect it does, however most of the rest of Kodi is Sentence case so maybe using that for our summaries is better.

KOPRajs replied:
We do have another translated string (#30000) that is being used in the port dialog and the controller dialog:
It is "Genesis (3-Button)" in this case.

When I search the UI for the summary I can only see it in the addon information dialog, nowhere else.
So the sentence case for the summary is fine, there is no problem with that.

The question is only about the content in the brackets in summary. Currently all controllers, which have anything in brackets in the summary, have capital letter.

I'm not native english speaker, so I'm not sure whether e.g. "_(european)_" is right (it is in my language). If that needs the capital, then only the "(Player 2)" and the "(4-Button)" are in question here.

garbear commented on addons/game.controller.genesis.mouse/addon.xml:
```diff
@@ -1,7 +1,7 @@
 <addon id="game.controller.genesis.mouse"
-        name="Sega Mouse"
-        version="1.0.21"
+        name="Genesis Mouse"
```
Why not "Sega Genesis Mouse" to improve the sort order?

KOPRajs replied:
It is a bit tricky here. If we want to match the "layout" then there is no reason to have "Sega" in the name. Other controllers do not have a manufacturer there as well.

We have:
Dreamcast Controller
Genesis Controller (3-Button)
Genesis Controller (6-Button)
Genesis Mouse
NES Controller
PlayStation Gamepad
PlayStation Mouse
SNES Controller
SNES Mouse

There's no _Nintendo_ (even if it is included in the console name itself), there's no _Sony_, so why _Sega_? Also there might be other Sega console with mouse support, so the original name _Sega Mouse_ is ambiguous. The sort order seems actually better now, because you have all Genesis controllers together.

But on the other hand, there is a bunch of other Sega controllers that currently do have _Sega_ in their name. I think we should change them to console name as well.

And also it seems that the product trademark here is actually "Sega Mouse", that is why I left it in the description.

I've decided to go with the _Genesis Mouse_ to match the layout, but I can change that if you don't like it.

KOPRajs replied:
Or is there any controller which can be used cross system?

Is it possible to connect the "Sega Mouse" to e.g. Dreamcast? Do we use the same controller profile in such case? If so, then I would stay with the original name (_Sega Mouse_).

garbear replied:
Genesis Mouse is good. I forgot that "Sega Genesis Controller (3-Button)" would be way too long, we need short gui strings.

garbear commented:
Picture looks great. There's no real copyright reason to attribute the image; all of this is clearly fair use. I attribute images as a favor to the creator. We do take a cost for image attribution, because it's a translatable string. So weight the costs/benefits, and I think it's better to not attribute the image here.

KOPRajs commented:
> I think that DefaultControllerTranslator.cpp uses the ergonomic mapping, is this true?

Yes, it seems so. One more reason for changing the buttonmaps of affected emulators (like NES and SNES) to match this.

> Search https://github.com/kodi-game/game.libretro/tree/Nexus/src/input for the tables the define the default fallbacks for the known controllers.

I think that I'm still missing one piece from the input puzzle. I've installed the new 3 button controller and I've added it to the topology.xml and the buttonmap.xml of an emulator for testing. I can now select the controller, I can map it and it works as expected, but I've had to manually assign the buttons in the Controls menu.

So my button map in _userdata/addon_data/peripheral.joystick/resources/buttonmaps/xml/linux/GamepadPlus_90b_9a.xml_ now looks like this:
```
<?xml version="1.0" ?>
<buttonmap>
    <device name="GamepadPlus" provider="linux" buttoncount="90" axiscount="9">
        <configuration />
        <controller id="game.controller.default">
            <feature name="a" button="0" />
            <feature name="b" button="1" />
            <feature name="back" button="10" />
            <feature name="down" axis="+7" />
            <feature name="left" axis="-6" />
            <feature name="leftbumper" button="6" />
            <feature name="leftstick">
                <up axis="-1" />
                <down axis="+1" />
                <right axis="+0" />
                <left axis="-0" />
            </feature>
            <feature name="leftthumb" button="13" />
            <feature name="lefttrigger" button="8" />
            <feature name="right" axis="+6" />
            <feature name="rightbumper" button="7" />
            <feature name="rightstick">
                <up axis="-3" />
                <down axis="+3" />
                <right axis="+2" />
                <left axis="-2" />
            </feature>
            <feature name="rightthumb" button="14" />
            <feature name="righttrigger" button="9" />
            <feature name="start" button="11" />
            <feature name="up" axis="-7" />
            <feature name="x" button="3" />
            <feature name="y" button="4" />
        </controller>
        <controller id="game.controller.genesis.3button">
            <feature name="a" button="3" />
            <feature name="b" button="0" />
            <feature name="c" button="1" />
            <feature name="down" axis="+7" />
            <feature name="left" axis="-6" />
            <feature name="mode" button="10" />
            <feature name="right" axis="+6" />
            <feature name="start" button="11" />
            <feature name="up" axis="-7" />
        </controller>
    </device>
</buttonmap>
```
The 6 button works without the need to manually assign the buttons but the 3 button does not. So there has to be some fallback mapping from _game.controller.default_ to _game.controller.genesis.6button_, which I can't find. This mapping somehow does work for the 6 button, but it doesn't work for the 3 button.

garbear commented:
I think your solution lies in https://github.com/xbmc/peripheral.joystick/tree/Nexus/src/buttonmapper

When a controller is connected, peripheral.joystick loads _all_ button maps. If any button map can provide a mapping between two controllers in either direction, it is used. If there are multiple mappings to chose from, they're sorted and the most popular map is chosen.

KOPRajs commented:
> When a controller is connected, peripheral.joystick loads _all_ button maps. If any button map can provide a mapping between two controllers in either direction, it is used. If there are multiple mappings to chose from, they're sorted and the most popular map is chosen.

I can see, finally it makes sense.

So the new _game.controller.genesis.3button_ doesn't work out-of-box, because there is no mapping for it in _addons/peripheral.joystick/resources/buttonmaps/xml/_.

In such case, this PR may be considered complete. If you agree with the cosmetics (uppercase X lowercase etc.), then let's merge this one.

Also this is the answer to "_Where should we adjust the default mapping for NES and SNES?_"  We need to make sure that the ergonomic mapping ("A"->"B", "X"->"Y") is "_the most popular_" one in _peripheral.joystick_. If we try to change it anywhere else, it can be always overridden by the maps in _peripheral.joystick_ anyway. And even worse, it depends on how many times the non-ergonomic mapping is present. And probably it is platform specific too (it depends on the used joystick driver).

garbear commented:
The initial goal of the input manager was: "Wrong input is better than no input". I've seen people plug in a controller, have it do nothing, and give up. I've also seen people plug in a controller, have it be slightly different from what they expect, and happily learn to play with the new configuration.

However, I think we've now passed the critical point where most controllers work for most games, and it's time to get most controllers correct for most games.

I actually planned for this inflection point, and it can be achieved without code changes and only XML changes. Simply remove multiple controllers from all button maps but one. One vote wins in a popularity of one. The single button map becomes the codex that all other controllers will use.

However, it's probably best to change the code to hard-code the codex approach. If you can take any controller and map every joystick profile in the UI, then we put that map in a special directory, and modify the code to only load that single map instead of every map.

(It's worth mentioning that loading a single map instead of every map will cause controllers to start working much faster in the UI. Currently there's a 1 or 2 second delay while the button maps load and parse until the first controller is responsive in the UI. I accepted the small delay because I assumed we'd move toward a single-codex approach eventually.)

KOPRajs commented:
> However, it's probably best to change the code to hard-code the codex approach.

I agree. Even better if this approach makes the controllers work faster! It is no problem when you are connecting only one controller, but when I'm trying to connect 4 controllers to play the Bomberman, it sometimes takes long (the delay is sometimes longer than 1-2 seconds).

However I can only provide the default maps for the controllers that I use (NES, SNES, Genesis and PSX).

garbear commented:
That's not a problem, it's easy to match the ergonomics of most controllers because a large picture is shown when you're mapping. For when we're unsure, just check the game add-on's buttonmap.xml and use whatever's there.

KOPRajs commented:
Ok, I will create the default mapping for all controller profiles using the latest Retroplayer release (there should be all controller profiles included in the build).

I've discussed the "Sentence case" vs. "Capital" in the brackets problem and the conclusion is to keep the capital in the brackets. So I'm going to rebase this to solve the version conflict, I'll add one more fix for the typo in the "Sega Virtua Gun (Eurpoean)" and then lets merge this before another version bump will break it again.

Lets continue the discussion about the default buttonmaps here: https://github.com/xbmc/peripheral.joystick/issues/259

KOPRajs commented:
@garbear I've fixed the conflict and added the typo fix, ready to go.

garbear commented:
Done!

garbear commented:
@KOPRajs The language CI workflow just sent some PRs to sync strings, can you see if everything looks right?

KOPRajs commented:
@garbear It looks like this commit has added line breaks: https://github.com/kodi-game/controller-topology-project/commit/d27e77eb58a5575457f2b1cdc7fd21d59071527d
But I don't understand why or even if anything in my PR have triggered it.
I've checked the git diff of this PR and I don't see anything wrong.

KOPRajs commented:
@garbear I don't know what exactly is this supposed to do, but isn't this automated commit broken?
https://github.com/kodi-game/controller-topology-project/commit/d27e77eb58a5575457f2b1cdc7fd21d59071527d

The same also happened last year here:
https://github.com/kodi-game/controller-topology-project/commit/75e7018fe6fbce175f3a5cd303af8b1af6dca0e4

The long one line strings are being broken into separate strings.

garbear commented:
Yup, something went wrong, I'm looking into it with @gade01

gade01 commented:
@garbear Is something wrong here?
It looks like Weblate is wrapping lines at 77 characters, which is fine.
Weblate will change it again at some point, as I configured it to only wrap at new lines.

garbear commented:
I think we're good. You can see I reverted a problematic commit in the history. I'll keep an eye on the repo just in case.

gade01 commented:
Yeah, I think so too. Thanks for fixing it.
It's a good idea to keep an eye on the repo.

Also, we don't want to lose any translations and cause frustrations for translators.
